src/snowflake/snowpark/mock/pandas_util.py

```diff
@@ -0,0 +1,214 @@
+                    # we need to calculate the precision and scale
+                    decimal_str = str(plain_data[row_idx][col_idx])
+                    decimal_parts = decimal_str.split(".")
+                    integer_len = (
+                        len(decimal_str)
+                        if len(decimal_parts) == 1
+                        else len(decimal_parts[0])
+                    )
+                    scale = 0 if len(decimal_parts) == 1 else len(decimal_parts[1])
+                    precision = integer_len + scale
+                    if precision > 38:
+                        raise SnowparkClientException(
+                            f"[Local Testing] Column precision {precision} and scale {scale} are not supported."
+                        )
+                    # handle integer and float separately
+                    data_type = DecimalType(precision=precision, scale=scale)
+                if previous_inferred_type:
+                    if isinstance(previous_inferred_type, NullType):
+                        inferred_type_dict[col_idx] = data_type
+                    if type(data_type) != type(previous_inferred_type):
+                        raise SnowparkClientException(
+                            "[Local Testing] Column contains different type of data."
+                        )
+                    if isinstance(inferred_type_dict, DecimalType):
+                        inferred_type_dict[col_idx] = DecimalType(
+                            precision=max(
+                                previous_inferred_type.precision, data_type.precision
+                            ),
+                            scale=max(previous_inferred_type.scale, data_type.scale),
+                        )
+                else:
+                    inferred_type_dict[col_idx] = data_type
+
+    fields = []
+    for idx, pandas_type in enumerate(data.dtypes):
+        if isinstance(pandas_type, pd.IntervalDtype):
+            data_type = VariantType()
+        elif isinstance(pandas_type, pd.DatetimeTZDtype):
+            data_type = TimestampType()
+        elif pandas_type.type == numpy.float64:
+            data_type = DoubleType()
+        elif isinstance(pandas_type, (pd.Float32Dtype, pd.Float64Dtype)):
+            data_type = DoubleType()
+        elif (
+            pandas_type.type == numpy.int64
+            or pandas_type.type == numpy.datetime64
+            or pandas_type.type == numpy.timedelta64
+        ):
+            data_type = LongType()
+        elif isinstance(pandas_type, PANDAS_INTEGER_TYPES):
+            data_type = LongType()
+        elif isinstance(pandas_type, pd.PeriodDtype):
+            data_type = LongType()
+        elif pandas_type.type == numpy.bool_:
+            data_type = BooleanType()
+        else:
+            data_type = inferred_type_dict.get(idx, StringType(length=16777216))
+        struct_field = StructField(col_names[idx], datatype=data_type, nullable=True)
```

Comment:
Question: does write_pandas always produce nullable columns?

Reply:
Let's check with `dtype="Int64"`

Reply:
tested that write_pandas always set nullable to True regardless of the dtype and the real data